Zone5/Transport/Zone5HTTPClient.swift

```diff
@@ -243,59 +268,52 @@ final public class Zone5HTTPClient {
 			decoder.keyDecodingStrategy = .useDefaultKeys
 			
 			// create with no completion handler. This will force delegate to be used so that we can capture progress
-			let task = urlSession.downloadTask(with: urlRequest) { [ weak self ] (location, response, error) in
-				
+			return urlSession.downloadTask(with: urlRequest) { location, response, error in
 				if let error = error {
-					self?.complete(completion, with: .failure(.transportFailure(error)))
+					completion(.failure(.transportFailure(error)))
 					return
 				}
 				
 				guard let response = response as? HTTPURLResponse, let filename = response.suggestedFilename else {
-					self?.complete(completion, with: .failure(.unknown))
+					completion(.failure(.unknown))
 					return
 				}
 				
 				let cacheURL = Zone5HTTPClient.downloadsDirectory.appendingPathComponent(filename)
 				
 				if (200..<400).contains(response.statusCode) {
 					if let resources = try? cacheURL.resourceValues(forKeys:[.fileSizeKey]), resources.fileSize! > 0 {
-						// success case - the file is the download. Pass this through the async queue directly
-						// because we want to delete the file afterwards
-						Zone5HTTPClient.clientHandlerQueue.async {
-							defer { try? FileManager.default.removeItem(at: cacheURL) }
-							completion(.success(cacheURL))
-						}
+						// success case - the file is the download
+						completion(.success(cacheURL))
+						try? FileManager.default.removeItem(at: cacheURL)
```

Comment:
note that the removeItem must be done synchronously after calling the completion handler. At the moment you have changed the completion execution from async to sync to this is ok - but in that case the whole wrapped completion handler is pointless. If it goes back to being asynchronous beware this code will delete the file before the completion handler can do something with it.

Reply:
Resolved this by wrapping the completion handler before passing it into the `execute(on:with:_:)` method, which ensures the deletion (both here and in the upload handler) is done before the underlying operation is completed. Between this and the fact that uploads/downloads are limited to one a piece, there shouldn't be any chance of cache clashes any more: the cache file is already gone by the time the next task is even started.